File: hs/src/Day01.hs
module Day01 (run) where

import Data.List 

run :: IO ()
run = do 
    contents <- readFile "input/01/input.txt"
    print . foldl (+) 0
          . take 3
          . reverse
          . sort
          . parseLines
          . lines $ contents

parseLines :: [String] -> [Integer]
parseLines xs = pl xs [0]
    where 
        pl ("":ys) acc = pl ys $ (acc ++ [0])
        pl (y:ys) acc = pl ys $ ((init acc) ++ [(last acc) + (readInt y)])
        pl [] acc = acc

readInt :: String -> Integer
readInt = read

File: ex/day01.exs
{:ok, contents} = File.read("input/01/input.txt")
contents
  |> String.split("\n")
  |> Enum.chunk_by(fn x -> x == "" end)
  |> Enum.filter(fn x -> Enum.at(x, 0) != "" end)
  |> Enum.map(fn x -> Enum.map(x, fn y -> String.to_integer(y) end) end)
  |> Enum.map(fn x -> Enum.sum(x) end)
  |> Enum.sort
  |> Enum.reverse
  |> Enum.slice(0, 3)
  |> Enum.sum
  |> IO.puts
